src/lib/actions/AbstractRunAction.ts

```diff
@@ -0,0 +1,94 @@
+import {InputsResolver} from "../InputsResolver";
+import {EngineOptionsFactory} from "../EngineOptionsFactory";
+import {INTERNAL_ERROR_CODE} from "../../Constants";
+
+/**
+ * Abstract Action to share a common implementation behind the "run" and "run dfa" commands
+ */
+export abstract class AbstractRunAction implements Action {
+	protected readonly display: Display;
+	private readonly inputsResolver: InputsResolver;
+	private readonly runOptionsFactory: RunOptionsFactory;
+	private readonly engineOptionsFactory: EngineOptionsFactory;
+
+	protected constructor(display: Display, inputsResolver: InputsResolver, runOptionsFactory: RunOptionsFactory,
+							engineOptionsFactory: EngineOptionsFactory) {
+		this.display = display;
+		this.inputsResolver = inputsResolver;
+		this.runOptionsFactory = runOptionsFactory;
+		this.engineOptionsFactory = engineOptionsFactory;
+	}
+
+	async validateInputs(inputs: Inputs): Promise<void> {
+		const fh = new FileHandler();
+		// If there's a --projectdir flag, its entries must be non-glob paths pointing to existing directories.
+		if (inputs.projectdir) {
+			// TODO: MOVE AWAY FROM ALLOWING AN ARRAY OF DIRECTORIES HERE AND ERROR IF THERE IS MORE THAN ONE DIRECTORY
+			for (const dir of (inputs.projectdir as string[])) {
+				if (globby.hasMagic(dir)) {
+					throw new SfError(getMessage(BundleName.CommonRun, 'validations.projectdirCannotBeGlob'));
+				} else if (!(await fh.exists(dir))) {
+					throw new SfError(getMessage(BundleName.CommonRun, 'validations.projectdirMustExist'));
+				} else if (!(await fh.stats(dir)).isDirectory()) {
+					throw new SfError(getMessage(BundleName.CommonRun, 'validations.projectdirMustBeDir'));
+				}
+			}
+		}
+		// If the user explicitly specified both a format and an outfile, we need to do a bit of validation there.
+		if (inputs.format && inputs.outfile) {
+			const inferredOutfileFormat: OUTPUT_FORMAT = inferFormatFromOutfile(inputs.outfile as string);
+			// For the purposes of this validation, we treat junit as xml.
+			const chosenFormat: string = inputs.format === 'junit' ? 'xml' : inputs.format as string;
+			// If the chosen format is TABLE, we immediately need to exit. There's no way to sensibly write the output
+			// of TABLE to a file.
+			if (chosenFormat === OUTPUT_FORMAT.TABLE) {
+				throw new SfError(getMessage(BundleName.CommonRun, 'validations.cannotWriteTableToFile', []));
+			}
+			// Otherwise, we want to be liberal with the user. If the chosen format doesn't match the outfile's extension,
+			// just log a message saying so.
+			if (chosenFormat !== inferredOutfileFormat) {
+				this.display.displayInfo(getMessage(BundleName.CommonRun, 'validations.outfileFormatMismatch', [inputs.format as string, inferredOutfileFormat]));
+			}
+		}
+	}
+
+	async run(inputs: Inputs): Promise<AnyJson> {
+		const filters: RuleFilter[] = new RuleFilterFactoryImpl().createRuleFilters(inputs);
```

Comment:
Question: The other classes that use a filter factory accept it as a parameter to the constructor. Is there a reason that's not done here?

Reply:
This was an oversite on my part. Good catch - will fix.

Reply:
Fixed - squashed into part 10 with rebase and then force push.